pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -449,64 +473,114 @@ impl ImageLayerInner {
 
-                        if !range.contains(&key) {
-                            return false;
+                        if key >= range.end {
+                            planner.handle_range_end(offset);
+                            range_end_handled = true;
+                            false
+                        } else {
+                            planner.handle(key, self.lsn, offset, BlobFlag::None);
+                            true
                         }
-
-                        offsets.push((key, value));
-
-                        true
                     },
                     &RequestContextBuilder::extend(ctx)
                         .page_content_kind(PageContentKind::ImageLayerBtreeNode)
                         .build(),
                 )
                 .await
                 .map_err(|err| GetVectoredError::Other(anyhow!(err)))?;
-        }
 
-        let ctx = &RequestContextBuilder::extend(ctx)
-            .page_content_kind(PageContentKind::ImageLayerValue)
-            .build();
-
-        let cursor = file.block_cursor();
-        let mut buf = Vec::new();
-        for (key, offset) in offsets {
-            let res = cursor.read_blob_into_buf(offset, &mut buf, ctx).await;
-            if let Err(e) = res {
-                reconstruct_state.on_key_error(
-                    key,
-                    PageReconstructError::from(anyhow!(e).context(format!(
-                        "Failed to read blob from virtual file {}",
-                        file.file.path
-                    ))),
-                );
-
-                continue;
+            if !range_end_handled {
+                let payload_end = self.index_start_blk as u64 * PAGE_SZ as u64;
+                planner.handle_range_end(payload_end);
             }
-
-            let blob = Bytes::copy_from_slice(buf.as_slice());
-            reconstruct_state.update_key(&key, self.lsn, Value::Image(blob));
         }
 
-        Ok(())
+        Ok(planner.finish())
+    }
+
+    async fn do_reads_and_update_state(
+        &self,
+        reads: Vec<VectoredRead>,
+        reconstruct_state: &mut ValuesReconstructState,
+    ) {
+        let mut buf = Some(BytesMut::with_capacity(
+            self.vectored_blob_reader.get_max_read_size(),
+        ));
+        for read in reads.into_iter().rev() {
+            let res = self
+                .vectored_blob_reader
+                .read_blobs(&read, buf.take().expect("Should have a buffer"))
+                .await;
+
+            match res {
+                Ok(blobs_buf) => {
+                    for meta in blobs_buf.blobs.iter().rev() {
+                        let img_buf = Bytes::copy_from_slice(&blobs_buf.buf[meta.start..meta.end]);
+                        reconstruct_state.update_key(
+                            &meta.meta.key,
+                            self.lsn,
+                            Value::Image(img_buf),
+                        );
```

Comment:
Please apply the same changes to `delta_layer` if applicable

Reply:
Ahh, no, the `blob_buf.buf` is being recycled for the next iteration.
Well, we can still save ourselves the `copy_from_slice` by not recycling the `blob_buf`, and instead allocating a new `buf` every iteration.

And while at it, we can make the `read_blobs` allocate the `blob_buf`, simplifies this function and the `read_blobs()` API

Reply:
> Well, we can still save ourselves the copy_from_slice by not recycling the blob_buf, and instead allocating a new buf every iteration.

Agreed, we can do this.

> And while at it, we can make the read_blobs allocate the blob_buf, simplifies this function and the read_blobs() API

Disagree because this means that we can't reuse the delta layer. We don't have to copy there.

Reply:
> Well, we can still save ourselves the copy_from_slice by not recycling the blob_buf, and instead allocating a new buf every iteration.

Done in https://github.com/neondatabase/neon/pull/6780/commits/73cbe4f5cf09a977c38fe456e94f47aab12c3281.

Won't resolve. Let me know if you disagree with my disagreement.

Reply:
Looks good. I'd still like to re-use the buffer at all times, seems like the consistent thing to do IMO.

Here's a patch. It's immaterial because error case, but, if there's a chance for buffer re-use, we should take it. Will simplify future work / experiments with using a buffer pool instead of ad-hoc allocations.

[reuse-buffer.patch](https://github.com/neondatabase/neon/files/14417806/reuse-buffer.patch)

Reply:
Follow up created: https://github.com/neondatabase/neon/issues/6932